### PR TITLE
FIX-1334: transform_reduce

### PR DESCRIPTION
**Pull request by DenisYaroshevskiy**
One algorithm to be used as an alternative to `algo::reduce(views::map)`
Because it allows for non template mappin operation, had been done separately.

### CODE REVIEW

File: include/eve/algo/reduce.hpp [hunk 1]
```diff
@@ -20,6 +20,40 @@
 
 namespace eve::algo
 {
+  //================================================================================================
+  //! @addtogroup algorithms
+  //! @{
+  //!  @var reduce
+  //!
+  //!  @brief SIMD version of std::reduce
```
Comment (jfalcou) [inline include/eve/algo/reduce.hpp:28]: I would appreciate if the new documentation for algo follows the schema from the copy.hpp file with respect to :
 - sub header
 - global location for information
 - use of doxygen @note or @warning accordingly

 It'll svae time later down the road.

File: include/eve/algo/transform_reduce.hpp [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+#include <utility>
+
+namespace eve::algo
+{
+//================================================================================================
+//! @addtogroup algorithms
+//! @{
+//!  @var transform_reduce
+//!
+//!  @brief SIMD version of std::transform_reduce for a single range
```
Comment (jfalcou) [inline include/eve/algo/transform_reduce.hpp:28]: Same than for reduce

File: test/unit/algo/algorithm/reduce_generic.h [hunk 3]
```diff
@@ -0,0 +1,115 @@
+//==================================================================================================
```
Comment (jfalcou) [inline test/unit/algo/algorithm/reduce_generic.h:1]: Why not .hpp like usual ?

File: test/unit/algo/algorithm/reduce_generic.h [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  {
+    auto alg = basic_reduce[eve::algo::unroll<1>][eve::algo::force_cardinal<T::size()>];
+    reduce_generic_test_page_ends(tgt, e_t{10}, alg);
+  }
+
+  // unroll 3
+  {
+    reduce_generic_test_page_ends(native_tgt, e_t{10}, basic_reduce[eve::algo::unroll<3>]);
+  }
+}
```
Comment (jfalcou) [inline test/unit/algo/algorithm/reduce_generic.h:115]: Could we add test for stateful reduction operation ?

File: test/unit/algo/algorithm/sums_special_cases.cpp [hunk 5]
```diff
@@ -69,21 +66,57 @@ TTS_CASE("eve.algo.inclusive_scan complex numbers")
...
   TTS_EQUAL(expected_real, real);
-  TTS_EQUAL(expected_img,  img);
+  TTS_EQUAL(expected_img, img);
 
   TTS_EQUAL(expected_real, real_copy);
-  TTS_EQUAL(expected_img,  img_copy);
+  TTS_EQUAL(expected_img, img_copy);
+};
+
+TTS_CASE("eve.algo.transform_reduce_simple")
```
Comment (jfalcou) [inline test/unit/algo/algorithm/sums_special_cases.cpp:88]: or should the stateful reduction happen here?

File: test/doc/algo/reduce.cpp [hunk 6]
```diff
@@ -1,5 +1,6 @@
 #include <eve/module/core.hpp>
 #include <eve/algo.hpp>
+#include <numeric>
 #include <iostream>
 #include <vector>
 #include "print.hpp"
```
Comment (jfalcou) [inline test/doc/algo/reduce.cpp:6]: due to how doc tests are able to be opened in Godbolt, we should not use print.hpp in those doc tests.
What I think is OK is to reproduce of the std cppreference samples, rise the size of the data and be happy.

Again, check how copy.cpp has been modified

File: test/doc/algo/transform_reduce.cpp [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+#include <numeric>
+#include <vector>
+#include "print.hpp"
+
+int main()
+{
+  std::vector<float>  v = {1.0f, 2.0f, 3.0f, 4.0f};
+
+  std::cout << " -> v                                                                                               = ";
+  doc_utils::print(v);
```
Comment (jfalcou) [inline test/doc/algo/transform_reduce.cpp:13]: Same. Avoid print

File: test/unit/algo/algorithm/sums_special_cases.cpp (hunk 5 above)
Comment (DenisYaroshevskiy) [reply test/unit/algo/algorithm/sums_special_cases.cpp:88]: I don't know a good example on stateful reduction. Stateful map added.

File: include/eve/algo/reduce.hpp (hunk 1 above)
Comment (DenisYaroshevskiy) [reply include/eve/algo/reduce.hpp]: Please recheck.

File: test/unit/algo/algorithm/sums_special_cases.cpp [hunk 8]
```diff
@@ -69,21 +66,91 @@ TTS_CASE("eve.algo.inclusive_scan complex numbers")
...
+
+TTS_CASE("eve.algo.reduce, stateful reduce")
+{
+  std::vector v {1, 2, 3};
+
+  TTS_EQUAL(6,
+            eve::algo::reduce(v,
+                              std::pair {[_ = 3](auto x, auto y)
+                                         {
+                                           (void)_;
```
Comment (jfalcou) [inline test/unit/algo/algorithm/sums_special_cases.cpp:132]:
```
[[maybe_unused]]_;
```

maybe ?

Comment (DenisYaroshevskiy) [reply test/unit/algo/algorithm/sums_special_cases.cpp:132]: doesn't work

**Review by jfalcou — APPROVED** [review]

**Pull request merged**